EventSourcing.Core/Aggregates/AggregateService.cs

```diff
@@ -28,101 +28,87 @@ public class AggregateService<TBaseEvent> : IAggregateService<TBaseEvent> where
+      .Where(x => x.PartitionId == partitionId && x.AggregateId == aggregateId && x.Timestamp <= date)
       .OrderBy(x => x.AggregateVersion)
-      .ToAsyncEnumerable();
+      .AsAsyncEnumerable();
       
-    return await Aggregate<TBaseEvent>.RehydrateAsync<TAggregate>(aggregateId, events, cancellationToken);
+    return await Aggregate<TBaseEvent>.RehydrateAsync<TAggregate>(partitionId, aggregateId, events, cancellationToken);
   }
     
-  private async Task<TAggregate> RehydrateFromSnapshotAsync<TAggregate>(Guid aggregateId,
+  private async Task<TAggregate> RehydrateFromSnapshotAsync<TAggregate>(Guid partitionId, Guid aggregateId,
     CancellationToken cancellationToken = default) where TAggregate : Aggregate<TBaseEvent>, new()
   {
     if (_snapshotStore == null)
       throw new InvalidOperationException("Snapshot store not provided");
       
-    var latestSnapshot = _snapshotStore.Snapshots
-      .Where(x => x.AggregateId == aggregateId)
+    var latestSnapshot = await _snapshotStore.Snapshots
+      .Where(x => x.PartitionId == partitionId && x.AggregateId == aggregateId)
       .OrderBy(x => x.AggregateVersion)
-      .LastOrDefault();
-      
+      .AsAsyncEnumerable()
+      .LastOrDefaultAsync(cancellationToken);
+
+    var version = latestSnapshot?.AggregateVersion ?? 0;
+
     var events = _eventStore.Events
-      .Where(x => x.AggregateId == aggregateId);
+      .Where(x => x.PartitionId == partitionId && x.AggregateId == aggregateId && x.AggregateVersion >= version)
+      .OrderBy(x => x.AggregateVersion)
+      .AsAsyncEnumerable();
 
     if (latestSnapshot != null)
-      events = events
-        .Where(x => x.AggregateVersion >= latestSnapshot.AggregateVersion)
-        .Prepend(latestSnapshot);
+      events = events.Prepend(latestSnapshot);
 
-    var orderedEvents = events.OrderBy(x => x.AggregateVersion).ToAsyncEnumerable();
-    return await Aggregate<TBaseEvent>.RehydrateAsync<TAggregate>(aggregateId, orderedEvents, cancellationToken);
+    return await Aggregate<TBaseEvent>.RehydrateAsync<TAggregate>(partitionId, aggregateId, events, cancellationToken);
   }
 
   public async Task<TAggregate> PersistAsync<TAggregate>(TAggregate aggregate,
     CancellationToken cancellationToken = default) where TAggregate : Aggregate<TBaseEvent>, new()
   {
-    if (aggregate.Id == Guid.Empty)
-      throw new ArgumentException("Aggregate.Id cannot be empty", nameof(aggregate));
-
-    await _eventStore.AddAsync(aggregate.UncommittedEvents.ToList(), cancellationToken);
+    var transaction = CreateTransaction(aggregate.PartitionId);
+    var result = await transaction.PersistAsync(aggregate, cancellationToken);
+    await transaction.CommitAsync(cancellationToken);
+    return result;
+  }
 
-    if (aggregate is ISnapshottable s && s.IntervalExceeded<TBaseEvent>())
+  public async Task PersistAsync<TAggregate>(IEnumerable<TAggregate> aggregates, CancellationToken cancellationToken = default)
```

Comment:
This method is very useful but I think it would even be better if we can remove the type parameter here since in some situations we need to able to store multiple types of aggregates at once.

Reply:
For sure!